src/particles/PhysicsParticles.hpp

```diff
@@ -0,0 +1,206 @@
+							    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> const &plo,
+							    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> const &dxi) const noexcept
+	{
+		amrex::ParticleInterpolator::Linear interp(p, plo, dxi);
+		interp.ParticleToMesh(p, rho, start_part_comp, start_mesh_comp, num_comp, [=] AMREX_GPU_DEVICE(const ParticleType &part, int comp) {
+			return 4.0 * M_PI * Gconst * part.rdata(comp); // weight by 4 pi G
+		});
+	}
+};
+
+struct RadDeposition {
+	double current_time{};
+	int start_part_comp{};
+	int start_mesh_comp{};
+	int num_comp{};
+
+	template <typename ParticleType>
+	AMREX_GPU_DEVICE AMREX_FORCE_INLINE void operator()(const ParticleType &p, amrex::Array4<amrex::Real> const &radEnergySource,
+							    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> const &plo,
+							    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> const &dxi) const noexcept
+	{
+		amrex::ParticleInterpolator::Linear interp(p, plo, dxi);
+		interp.ParticleToMesh(p, radEnergySource, start_part_comp, start_mesh_comp, num_comp, [=] AMREX_GPU_DEVICE(const ParticleType &part, int comp) {
+			if (current_time < part.rdata(RadParticleBirthTimeIdx) || current_time >= part.rdata(RadParticleDeathTimeIdx)) {
+				return 0.0;
+			}
+			return part.rdata(comp) * (AMREX_D_TERM(dxi[0], *dxi[1], *dxi[2]));
+		});
+	}
+};
+
+// Forward declarations
+template <typename problem_t> class PhysicsParticleRegister;
+
+// Base class for physics particle descriptors
+class PhysicsParticleDescriptor
+{
+      protected:
+	int massIndex_{-1};		 // index for gravity mass, -1 if not used
+	int lumIndex_{-1};		 // index for radiation luminosity, -1 if not used
+	bool interactsWithHydro_{false}; // whether particles interact with hydro
+
+      public:
+	PhysicsParticleDescriptor(int mass_idx, int lum_idx, bool hydro_interact)
+	    : massIndex_(mass_idx), lumIndex_(lum_idx), interactsWithHydro_(hydro_interact)
+	{
+	}
+	virtual ~PhysicsParticleDescriptor() = default;
+	void *neighborParticleContainer_{}; // pointer to particle container, type-erased
+
+	// Getters
+	[[nodiscard]] auto getMassIndex() const -> int { return massIndex_; }
+	[[nodiscard]] auto getLumIndex() const -> int { return lumIndex_; }
+	[[nodiscard]] auto getInteractsWithHydro() const -> bool { return interactsWithHydro_; }
+
+	// Virtual methods that can be overridden
+	virtual void hydroInteract() {} // Default no-op
+
+	// Delete copy/move constructors/assignments
+	PhysicsParticleDescriptor(const PhysicsParticleDescriptor &) = delete;
+	PhysicsParticleDescriptor &operator=(const PhysicsParticleDescriptor &) = delete;
+	PhysicsParticleDescriptor(PhysicsParticleDescriptor &&) = delete;
+	PhysicsParticleDescriptor &operator=(PhysicsParticleDescriptor &&) = delete;
+};
+
+// Registry for physics particles
+template <typename problem_t> class PhysicsParticleRegister
+{
+      private:
+	std::map<std::string, std::unique_ptr<PhysicsParticleDescriptor>> particleRegistry_;
```

Comment:
particleRegistry_ has to be a map from string to std::unique_ptr because it's used to store a descriptor, which is a `std::unique_ptr<ParticleDescriptor>`

Reply:
Ok, I don't follow. Why do the descriptors need to be `std::unique_ptr<ParticleDescriptor>` instead of just `ParticleDescriptor`?

Reply:
I believe changing it to `std::map<std::string, PhysicsParticleDescriptor>` would make a copy of PhysicsParticleDescriptor, which should be avoided. 

In simulation.hpp, radParticleDesc is created as a pointer, so it makes sense to define `particleRegistry_` as a map from string to a pointer. 
```cpp
		auto radParticleDesc = std::make_unique<quokka::PhysicsParticleDescriptor>(-1, quokka::RadParticleLumIdx, quokka::RadParticleBirthTimeIdx, false);
```

Reply:
You don't have to create it as a pointer. If you then `std::move` it, it won't be copied.